velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -28,52 +28,177 @@ namespace facebook::velox::exec::prefixsort {
+/// integer, their sign-bit are same, flip sign-bit do not change result.
+/// Compare two negative signed integers: -n = ~n + 1, we can trait ~n + 1 as an
+/// unsigned integer, so the logic is as same as unsigned integer, also flip
+/// sign-bit do not change result.
+/// Compare positive vs negative: flip sign-bit to promise that positive
+/// always bigger than negative.
 template <>
-FOLLY_ALWAYS_INLINE void PrefixSortEncoder::decode<int64_t>(char* row) {
-  row[0] = flipSignBit(row[0]);
-  const auto v = __builtin_bswap64(*reinterpret_cast<uint64_t*>(row));
-  simd::memcpy(row, &v, sizeof(int64_t));
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    int32_t value,
+    char* dest) const {
+  encodeNoNulls((uint32_t)(value ^ 1u << 31), dest);
 }
 
-/// For testing only.
-namespace {
-template <typename T>
-FOLLY_ALWAYS_INLINE void testingEncodeInPlace(const std::vector<T>& data) {
-  for (auto i = 0; i < data.size(); i++) {
-    PrefixSortEncoder::encode(data[i], (char*)data.data() + i * sizeof(T));
+/// Logic is nearly as same as int32_t
+template <>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    uint64_t value,
+    char* dest) const {
+  auto& v = *reinterpret_cast<uint64_t*>(dest);
+  v = __builtin_bswap64(value);
+  if (!ascending_) {
+    v = ~v;
   }
 }
 
-template <typename T>
-FOLLY_ALWAYS_INLINE void testingDecodeInPlace(const std::vector<T>& data) {
-  for (auto i = 0; i < data.size(); i++) {
-    PrefixSortEncoder::decode<T>((char*)data.data() + i * sizeof(T));
+template <>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    int64_t value,
+    char* dest) const {
+  encodeNoNulls((uint64_t)(value ^ 1ull << 63), dest);
+}
+
+/// Convert double to a uint64_t, their value comparison semantics remain
+/// consistent
+static FOLLY_ALWAYS_INLINE uint64_t encodeDouble(double x) {
+  uint64_t buff;
+  // Zero is the smallest positive value.
+  if (x == 0) {
+    return 1ull << 63;
+  }
+  // Nan is max value
+  if (std::isnan(x)) {
+    return ULLONG_MAX;
+  }
+  // Infinity is the second max value
+  if (x > std::numeric_limits<double>::max()) {
+    return ULLONG_MAX - 1;
   }
+  // -Infinity is the smallest value
+  if (x < -std::numeric_limits<double>::max()) {
+    return 0;
+  }
+  buff = *reinterpret_cast<uint64_t*>(&x);
+  if ((buff & (1ull << 63)) == 0) {
+    // For positive numbers, set sign bit to 1.
+    buff |= (1ull << 63);
+  } else {
+    // For negative numbers, invert bits to get the opposite order.
+    buff = ~buff;
+  }
+  return buff;
+}
+
+/// If the result of encodeDouble() keeps value comparison semantics, then we
+/// can trait it as an unsigned-integer.
+template <>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    double value,
+    char* dest) const {
+  encodeNoNulls(encodeDouble(value), dest);
+}
+
+// Logic is as same as double.
+static FOLLY_ALWAYS_INLINE uint32_t encodeFloat(float x) {
```

Comment:
Let's put both encodeDouble and encodeFloat into a "detail" namespace.

x -> value